File: tools/dropins/p2-set032-audit-event-writer/src/Admin/Migration.Admin.Web/src/api/auditEventWriter.ts
import { apiPost } from './core/adminApiClient';
import type { AuditWriteResult } from './auditPersistence';

export type AuditEventWriteRequest = {
  workspaceId: string;
  category: string;
  eventName: string;
  severity: string;
  tenantId?: string | null;
  projectId?: string | null;
  runId?: string | null;
  correlationId?: string | null;
  actor?: string | null;
  properties?: Record<string, string> | null;
};

export type AuditEventWriterProbe = {
  request: AuditEventWriteRequest;
  result: AuditWriteResult;
};

export async function probeAuditEventWriter(): Promise<AuditEventWriterProbe> {
  return apiPost<AuditEventWriterProbe>('/api/cloud/audit/writer/probe', {});
}

File: tools/dropins/p2-set032-audit-event-writer/src/Migration.ControlPlane/Audit/AuditEventWriter.cs
namespace Migration.ControlPlane.Audit;

public sealed class AuditEventWriter : IAuditEventWriter
{
    private readonly IAuditPersistenceProvider _provider;

    public AuditEventWriter(IAuditPersistenceProvider provider)
    {
        _provider = provider;
    }

    public Task<AuditWriteResult> WriteAsync(
        AuditEventWriteRequest request,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);

        var record = AuditRecordFactory.Create(
            workspaceId: request.WorkspaceId,
            category: request.Category,
            eventName: request.EventName,
            severity: request.Severity,
            tenantId: request.TenantId,
            projectId: request.ProjectId,
            runId: request.RunId,
            correlationId: request.CorrelationId,
            actor: request.Actor,
            properties: request.Properties);

        return _provider.WriteAsync(record, cancellationToken);
    }
}
